### DISCUSSION TITLE
blocking-primary-constructor-member-capture-using-an-analyzer/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#421] [Announcements]
# blocking-primary-constructor-member-capture-using-an-analyzer/

In this post I describe a Rosyln analyzer you can use to enforce that primary constructors are only used for initialization and never to capture as fields

https://andrewlock.net/blocking-primary-constructor-member-capture-using-an-analyzer/

**Comment by jhigh2000** [comment]
This analyzer is a great idea and addresses the main concern I have with primary constructors. Are there other steps that need to be taken to enable this analyzer? I've added the code above but can't seem to get it to trigger.

**Reply by TallBaldGeek** [reply]
Apologies if this is too basic of a reply, but [this blog post](https://devblogs.microsoft.com/dotnet/how-to-write-a-roslyn-analyzer/) and [this tutorial](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix) helped me get familiar with writing analyzers. Maybe they'll help?

**Comment by TallBaldGeek** [comment]
Hi Andrew - I've been doing some Roslyn analyzer work for a company project and we're running into issues in our CI/CD pipeline because the testing NuGet package `Microsoft.CodeAnalysis.Analyzer.Testing` hasn't had a release in a while and v1.1.1 causes the build to attempt a download of .NET Core 3.1 which is out of support. (See [this issue for details](https://github.com/dotnet/roslyn-sdk/issues/1047#issuecomment-1897506914))

I see you're using a similar process for testing your analyzers, and your [Datadog.Trace.Tools.Analyzers.Tests.csproj](https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/test/Datadog.Trace.Tools.Analyzers.Tests/Datadog.Trace.Tools.Analyzers.Tests.csproj) is still targeting net5.0 and v1.1.1 of the `Microsoft.CodeAnalysis.*` libraries. Have you encountered any issues as a result?

**Comment by bart-vmware** [comment]
Thanks, great article! For anyone using Resharper, a special setting exists to prefer readonly fields over implicit captures, see https://youtrack.jetbrains.com/issue/RSRP-494712/Redundant-members-cleanup-step-in-Convert-to-primary-constructor-action-changes-behavior-of-readonly-field.#focus=Comments-27-8525300.0-0.
